File: ComPDFKit for Windows/Examples/Samples/EncryptTest/VB/EncryptTest.vb
Imports ComPDFKit.PDFDocument
Imports System
Imports System.IO

Module EncryptTest
    Private outputPath As String = Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(System.IO.Directory.GetCurrentDirectory()))) & "\Output\VB"
    Private userPassword As String = String.Empty
    Private ownerPassword As String = String.Empty

    Sub Main(args As String())
        Console.WriteLine("Running Encrypt test sample..." & vbCrLf)
        SDKLicenseHelper.LicenseVerify()

        If Not Directory.Exists(outputPath) Then
            Directory.CreateDirectory(outputPath)
        End If

        ' Encrypt by user password
        Dim document As CPDFDocument = CPDFDocument.InitWithFilePath("CommonFivePage.pdf")

        If EncryptByUserPassword(document) Then
            Console.WriteLine("Encrypt by user password done.")
        Else
            Console.WriteLine("Encrypt by user password failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

        ' Encrypt by owner password
        document = CPDFDocument.InitWithFilePath("CommonFivePage.pdf")

        If EncryptByOwnerPassword(document) Then
            Console.WriteLine("Encrypt by owner password done.")
        Else
            Console.WriteLine("Encrypt by owner password failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

        ' Encrypt by both user and owner passwords
        document = CPDFDocument.InitWithFilePath("CommonFivePage.pdf")

        If EncryptByAllPasswords(document) Then
            Console.WriteLine("Encrypt by Both user and owner passwords done.")
        Else
            Console.WriteLine("Encrypt by Both user and owner passwords failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

        ' Unlock
        document = CPDFDocument.InitWithFilePath("AllPasswords.pdf")

        If Unlock(document) Then
            Console.WriteLine("Unlock done.")
        Else
            Console.WriteLine("Unlock failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

        ' Decrypt
        document = CPDFDocument.InitWithFilePath("AllPasswords.pdf")

        If Decrypt(document) Then
            Console.WriteLine("Decrypt done.")
        Else
            Console.WriteLine("Decrypt failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

        Console.WriteLine("Done!")
        Console.WriteLine("--------------------")
        Console.ReadLine()
    End Sub

    Private Function EncryptUseRC4Algo(document As CPDFDocument, permissionsInfo As CPDFPermissionsInfo) As Boolean
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptUseRC4Test.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted")
            Console.WriteLine("Browse the changed file in: " & encryptPath)
            Console.WriteLine("User password is: {0}", userPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Function EncryptUseAES128Algo(document As CPDFDocument, permissionsInfo As CPDFPermissionsInfo) As Boolean
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelAES128
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptUseAES128Test.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted")
            Console.WriteLine("Browse the changed file in: " & encryptPath)
            Console.WriteLine("User password is: {0}", userPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Function EncryptUseAES256Algo(document As CPDFDocument, permissionsInfo As CPDFPermissionsInfo) As Boolean
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelAES256
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptUseAES256Test.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted")
            Console.WriteLine("Browse the changed file in " & encryptPath)
            Console.WriteLine("User password is: {0}", userPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Function EncryptUseNoEncryptAlgo(document As CPDFDocument, permissionsInfo As CPDFPermissionsInfo) As Boolean
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelNoEncryptAlgo
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptUseNoEncryptAlgoTest.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted.")
            Console.WriteLine("Browse the changed file in " & encryptPath)
            Console.WriteLine("User password is: {0}", userPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Function EncryptByUserPassword(document As CPDFDocument) As Boolean
        Dim result As Boolean = True
        userPassword = "User"
        ownerPassword = String.Empty
        Dim permissionsInfo As New CPDFPermissionsInfo()

        If EncryptUseRC4Algo(document, permissionsInfo) Then
            Console.WriteLine("RC4 encrypt done." & vbCrLf)
        Else
            Console.WriteLine("RC4 encrypt failed." & vbCrLf)
            result = False
        End If

        If EncryptUseAES128Algo(document, permissionsInfo) Then
            Console.WriteLine("AES128 encrypt done." & vbCrLf)
        Else
            Console.WriteLine("AES128 encrypt failed." & vbCrLf)
            result = False
        End If

        If EncryptUseAES256Algo(document, permissionsInfo) Then
            Console.WriteLine("AES256 encrypt done." & vbCrLf)
        Else
            Console.WriteLine("AES256 encrypt failed." & vbCrLf)
            result = False
        End If

        If EncryptUseNoEncryptAlgo(document, permissionsInfo) Then
            Console.WriteLine("NoEncryptAlgo encrypt done." & vbCrLf)
        Else
            Console.WriteLine("NoEncryptAlgo encrypt failed." & vbCrLf)
            result = False
        End If

        Return result
    End Function

    Private Function EncryptByOwnerPassword(document As CPDFDocument) As Boolean
        userPassword = Nothing
        ownerPassword = "Owner"
        Dim permissionsInfo As New CPDFPermissionsInfo()
        permissionsInfo.AllowsPrinting = False
        permissionsInfo.AllowsCopying = False
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptByOwnerPasswordTest.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted.")
            Console.WriteLine("Browse the changed file in " & encryptPath)
            Console.WriteLine("Owner password is: {0}", ownerPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Function EncryptByAllPasswords(document As CPDFDocument) As Boolean
        userPassword = "User"
        ownerPassword = "Owner"
        Dim permissionsInfo As New CPDFPermissionsInfo()
        permissionsInfo.AllowsPrinting = False
        permissionsInfo.AllowsCopying = False
        Dim encryptionLevel As CPDFDocumentEncryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4
        document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel)
        Dim encryptPath As String = outputPath & "\EncryptByAllPasswordsTest.pdf"

        If Not document.WriteToFilePath(encryptPath) Then
            Return False
        End If

        Dim encryptedDoc As CPDFDocument = CPDFDocument.InitWithFilePath(encryptPath)

        If encryptedDoc.IsEncrypted Then
            Console.WriteLine("File is encrypted.")
            Console.WriteLine("Browse the changed file in " & encryptPath)
            Console.WriteLine("User password is: {0}", userPassword)
            Console.WriteLine("Owner password is: {0}", ownerPassword)
        Else
            Console.WriteLine("File encrypt failed")
            Return False
        End If

        Return True
    End Function

    Private Sub PrintPermissionsInfo(permissionsInfo As CPDFPermissionsInfo)
        Console.Write("AllowsPrinting: ")
        Console.Write(If(permissionsInfo.AllowsPrinting = True, "Yes" & vbCrLf, "No" & vbCrLf))
        Console.Write("AllowsCopying: ")
        Console.Write(If(permissionsInfo.AllowsCopying = True, "Yes" & vbCrLf, "No" & vbCrLf))
    End Sub

    Private Function Unlock(document As CPDFDocument) As Boolean
        userPassword = "User"
        ownerPassword = "Owner"

        If document.IsLocked Then
            Console.WriteLine("Document is locked")
        End If

        PrintPermissionsInfo(document.GetPermissionsInfo())

        Console.WriteLine("Unlock with owner password")
        document.CheckOwnerPassword("123")

        PrintPermissionsInfo(document.GetPermissionsInfo())

        Return True
    End Function

    Private Function Decrypt(document As CPDFDocument) As Boolean
        userPassword = "User"
        ownerPassword = "Owner"
        Dim decryptPath As String = outputPath & "\DecryptTest.pdf"
        document.UnlockWithPassword(userPassword)

        If Not document.Decrypt(decryptPath) Then
            Return False
        End If

        Dim decryptDocument As CPDFDocument = CPDFDocument.InitWithFilePath(decryptPath)

        If decryptDocument.IsEncrypted Then
            Return False
        Else
            Console.WriteLine("Document decrypt done.")
        End If

        Return True
    End Function
End Module

File: ComPDFKit for Windows/Examples/Samples/EncryptTest/CS/EncryptTest.cs
using ComPDFKit.PDFDocument;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EncryptTest
{
    internal class EncryptTest
    {
        private static string outputPath =Path.Combine(Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(System.IO.Directory.GetCurrentDirectory()))) ?? string.Empty, "Output", "CS");
        static private string userPassword = string.Empty;
        static private string ownerPassword = string.Empty;

        static void Main(string[] args)
        {
            #region Perparation work

            Console.WriteLine("Running Encrypt test sample…\r\n");

            SDKLicenseHelper.LicenseVerify();

            if (!Directory.Exists(outputPath))
            {
                Directory.CreateDirectory(outputPath);
            }

            #endregion

            #region Sample 1: Encrypt by user password

            CPDFDocument document = CPDFDocument.InitWithFilePath("CommonFivePage.pdf");

            if (EncryptByUserPassword(document))
            {
                Console.WriteLine("Encrypt by user password done.");
            }
            else
            {
                Console.WriteLine("Encrypt by user password failed.");
            }

            document.Release();
            Console.WriteLine("--------------------");

            #endregion

            #region Sample 2:  Encrypt by owner password

            document = CPDFDocument.InitWithFilePath("CommonFivePage.pdf");

            if (EncryptByOwnerPassword(document))
            {
                Console.WriteLine("Encrypt by owner password done.");
            }
            else
            {
                Console.WriteLine("Encrypt by owner password failed.");
            }
            document.Release();

            Console.WriteLine("--------------------");

            #endregion

            #region Sample 3: Encrypt by all passwords

            document = CPDFDocument.InitWithFilePath("CommonFivePage.pdf");

            if (EncryptByAllPasswords(document))
            {
                Console.WriteLine("Encrypt by Both user and owner passwords done.");
            }
            else
            {
                Console.WriteLine("Encrypt by Both user and owner passwords failed.");
            }

            document.Release();

            Console.WriteLine("--------------------");

            #endregion

            #region Sample 4: Unlock
            document = CPDFDocument.InitWithFilePath("AllPasswords.pdf");
            if (Unlock(document))
            {
                Console.WriteLine("Unlock done.");
            }
            else
            {
                Console.WriteLine("Unlock failed.");
            }
            document.Release();

            Console.WriteLine("--------------------");

            #endregion

            #region Sample 5: Decrypt

            document = CPDFDocument.InitWithFilePath("AllPasswords.pdf");
            if (Decrypt(document))
            {
                Console.WriteLine("Decrypt done.");
            }
            else
            {
                Console.WriteLine("Decrypt failed.");
            }
            document.Release();
            Console.WriteLine("--------------------");

            #endregion

            Console.WriteLine("Done!");
            Console.WriteLine("--------------------");
            Console.ReadLine();
        }

        static private bool EncryptUseRC4Algo(CPDFDocument document, CPDFPermissionsInfo permissionsInfo)
        {
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptUseRC4Test.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted");
                Console.WriteLine("Browse the changed file in: " + encryptPath);
                Console.WriteLine("User password is: {0}", userPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }

            return true;
        }

        static private bool EncryptUseAES128Algo(CPDFDocument document, CPDFPermissionsInfo permissionsInfo)
        {
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelAES128;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptUseAES128Test.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted");
                Console.WriteLine("Browse the changed file in: " + encryptPath);
                Console.WriteLine("User password is: {0}", userPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }

            return true;
        }

        static private bool EncryptUseAES256Algo(CPDFDocument document, CPDFPermissionsInfo permissionsInfo)
        {
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelAES256;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptUseAES256Test.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted");
                Console.WriteLine("Browse the changed file in " + encryptPath);
                Console.WriteLine("User password is: {0}", userPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }

            return true;
        }

        static private bool EncryptUseNoEncryptAlgo(CPDFDocument document, CPDFPermissionsInfo permissionsInfo)
        {
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelNoEncryptAlgo;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptUseNoEncryptAlgoTest.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted.");
                Console.WriteLine("Browse the changed file in " + encryptPath);
                Console.WriteLine("User password is: {0}", userPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }
            return true;
        }

        /// <summary>
        /// Using RC4, AES128, AES256, NoEncrypt algorithm to encrypt document. 
        /// User password: User
        /// No owner password
        /// </summary>
        /// <param name="document">Regular document</param>
        static private bool EncryptByUserPassword(CPDFDocument document)
        {
            bool result = true;

            userPassword = "User";
            ownerPassword = string.Empty;
            CPDFPermissionsInfo permissionsInfo = new CPDFPermissionsInfo();

            if (EncryptUseRC4Algo(document, permissionsInfo))
            {
                Console.WriteLine("RC4 encrypt done.\n");
            }
            else
            {
                Console.WriteLine("RC4 encrypt failed.\n");
                result = false;
            }

            if (EncryptUseAES128Algo(document, permissionsInfo))
            {
                Console.WriteLine("AES128 encrypt done.\n");
            }
            else
            {
                Console.WriteLine("AES128 encrypt failed.\n");
                result = false;
            }

            if (EncryptUseAES256Algo(document, permissionsInfo))
            {
                Console.WriteLine("AES256 encrypt done.\n");
            }
            else
            {
                Console.WriteLine("AES256 encrypt failed.\n");
                result = false;
            }

            if (EncryptUseNoEncryptAlgo(document, permissionsInfo))
            {
                Console.WriteLine("NoEncryptAlgo encrypt done.\n");
            }
            else
            {
                Console.WriteLine("NoEncryptAlgo encrypt failed.\n");
                result = false;
            }
            return result;
        }

        /// <summary>
        /// Encrypt by owner password
        /// No user password
        /// Owner password: Owner
        /// </summary>
        /// <param name="document">Regular document</param> 
        static private bool EncryptByOwnerPassword(CPDFDocument document)
        {
            userPassword = null;
            ownerPassword = "Owner";
            CPDFPermissionsInfo permissionsInfo = new CPDFPermissionsInfo();
            permissionsInfo.AllowsPrinting = false;
            permissionsInfo.AllowsCopying = false;
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptByOwnerPasswordTest.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted.");
                Console.WriteLine("Browse the changed file in " + encryptPath);
                Console.WriteLine("Owner password is: {0}", ownerPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }
            return true;
        }

        /// <summary>
        /// Encrypt by all passwords
        /// User password: User
        /// Owner password: Owner
        /// </summary>
        /// <param name="document">Regular document</param> 
        static private bool EncryptByAllPasswords(CPDFDocument document)
        {
            userPassword = "User";
            ownerPassword = "Owner";
            CPDFPermissionsInfo permissionsInfo = new CPDFPermissionsInfo();
            permissionsInfo.AllowsPrinting = false;
            permissionsInfo.AllowsCopying = false;
            CPDFDocumentEncryptionLevel encryptionLevel = CPDFDocumentEncryptionLevel.CPDFDocumentEncryptionLevelRC4;
            document.Encrypt(userPassword, ownerPassword, permissionsInfo, encryptionLevel);
            string encryptPath = Path.Combine(outputPath, "EncryptByAllPasswordsTest.pdf");
            if (!document.WriteToFilePath(encryptPath))
            {
                return false;
            }

            CPDFDocument encryptedDoc = CPDFDocument.InitWithFilePath(encryptPath);
            if (encryptedDoc.IsEncrypted)
            {
                Console.WriteLine("File is encrypted.");
                Console.WriteLine("Browse the changed file in " + encryptPath);
                Console.WriteLine("User password is: {0}", userPassword);
                Console.WriteLine("Owner password is: {0}", ownerPassword);
            }
            else
            {
                Console.WriteLine("File encrypt failed");
                return false;
            }
            return true;
        }

        /// <summary>
        /// Print printing and copying permissions info.
        /// </summary>
        /// <param name="permissionsInfo">Permissions info about a document</param>
        static private void PrintPermissionsInfo(CPDFPermissionsInfo permissionsInfo)
        {
            Console.Write("AllowsPrinting: ");
            Console.Write(permissionsInfo.AllowsPrinting == true ? "Yes\n" : "No\n");
            Console.Write("AllowsCopying: ");
            Console.Write(permissionsInfo.AllowsCopying == true ? "Yes\n" : "No\n");
        }

        /// <summary>
        /// Unlock document
        /// </summary>
        /// <param name="document"></param>
        static private bool Unlock(CPDFDocument document)
        {
            userPassword = "User";
            ownerPassword = "Owner";
            //Check if document is locked
            if (document.IsLocked)
            {
                Console.WriteLine("Document is locked");
            }
               
            PrintPermissionsInfo(document.GetPermissionsInfo());
             
            Console.WriteLine("Unlock with owner password");
            document.CheckOwnerPassword("123");
            //Check permissions info.
            PrintPermissionsInfo(document.GetPermissionsInfo());
            return true;
        }

        /// <summary>
        /// Decrypted
        /// </summary>
        /// <param name="document"></param>
        static private bool Decrypt(CPDFDocument document)
        {
            userPassword = "User";
            ownerPassword = "Owner";
            string decryptPath = Path.Combine(outputPath, "DecryptTest.pdf");
            document.UnlockWithPassword(userPassword);
            if (!document.Decrypt(decryptPath))
            {
                return false;
            }
            CPDFDocument decryptDocument = CPDFDocument.InitWithFilePath(decryptPath);
            if (decryptDocument.IsEncrypted)
            {
                return false;
            }
            else
            {
                Console.WriteLine("Document decrypt done.");
            } 
            return true;
        }
    }
}
